tfhe/src/shortint/ciphertext/compressed_ciphertext.rs

```diff
@@ -0,0 +1,171 @@
+    ) -> Ciphertext {
+        let ct = packed_ct.0.extract();
+
+        let mut output = LweCiphertext::from_container(
+            vec![0; self.ciphertext_lwe_dimension().to_lwe_size().0],
+            self.ciphertext_modulus,
+        );
+
+        ShortintEngine::with_thread_local_mut(|engine| {
+            let (mut ciphertext_buffers, buffers) = engine.get_buffers(self);
+            match self.pbs_order {
+                PBSOrder::KeyswitchBootstrap => {
+                    apply_bootstrap(
+                        &self.bootstrapping_key,
+                        &ct.as_view(),
+                        &mut output.as_mut_view(),
+                        acc,
+                        buffers,
+                    );
+                }
+                PBSOrder::BootstrapKeyswitch => {
+                    apply_bootstrap(
+                        &self.bootstrapping_key,
+                        &ct.as_view(),
+                        &mut ciphertext_buffers.buffer_lwe_after_pbs,
+                        acc,
+                        buffers,
+                    );
+
+                    keyswitch_lwe_ciphertext(
+                        &self.key_switching_key,
+                        &ciphertext_buffers.buffer_lwe_after_pbs,
+                        &mut output.as_mut_view(),
+                    );
+                }
+            }
+        });
+
+        Ciphertext::new(
+            output,
+            acc.degree,
+            NoiseLevel::NOMINAL,
+            self.message_modulus,
+            self.carry_modulus,
+            self.pbs_order,
+        )
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use rand::Rng;
+
+    use crate::shortint::keycache::KEY_CACHE;
+    use crate::shortint::parameters::PARAM_MESSAGE_2_CARRY_2_KS_PBS;
+    use crate::shortint::PBSParameters;
+
+    const NB_TESTS: usize = 10;
+
+    // create_parametrized_test!(shortint_message_extract);
+
+    #[test]
+    fn shortint_test_modulus_switch_compressions() {
+        shortint_modulus_switch_compression(PARAM_MESSAGE_2_CARRY_2_KS_PBS);
+    }
+
+    fn shortint_modulus_switch_compression<P>(param: P)
+    where
+        P: Into<PBSParameters>,
+    {
+        let keys = KEY_CACHE.get_from_param(param);
+        let (cks, sks) = (keys.client_key(), keys.server_key());
+
+        let mut rng = rand::thread_rng();
+
+        let modulus_sup =
+            (cks.parameters.message_modulus().0 * cks.parameters.carry_modulus().0) as u64;
+
+        let modulus = cks.parameters.message_modulus().0 as u64;
+
+        for _ in 0..NB_TESTS {
+            let clear = rng.gen::<u64>() % modulus_sup;
+
+            let ctxt = cks.unchecked_encrypt(clear);
+
+            let packed_ct = sks.switch_modulus_and_pack(&ctxt);
+
+            let lookup_table = sks.generate_msg_lookup_table(|a| a + 1, ctxt.message_modulus);
+
+            let unpacked_ct = sks.unpack_and_apply_lookup_table(&packed_ct, &lookup_table);
+
+            let dec = cks.decrypt(&unpacked_ct);
+
```

Comment:
check degrees of ctxt and unpacked ct are the same with a non programmable decompression which applies the identity lut just to refresh the noise

Reply:
Done